keps/sig-network/4969-cluster-domain-downward-api/README.md

```diff
@@ -0,0 +1,938 @@
+by the cluster administrator.
+
+Currently, there is no way for cluster workloads to query for this domain name,
+leaving them to either use relative domain names or configure it manually.
+
+This KEP proposes adding a new field mapping into the Downward API mechanism, that workloads can use to request the cluster domain name.
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Relative domain names are a source of ambiguity: does `get.app` refer to
+the domain registry [get.app](https://get.app/) or the Service `get` in `app`?
+This also becomes problematic for TLS, since there is no way to distinguish
+which of these two cases a certificate applies to.
+
+Fully Qualified Domain Names (FQDNs) can be used to resolve this, by always
+specifying the full domain name. However, requiring each workload to configure
+the cluster domain is tedious and error-prone, discouraging application 
+developers from using them.
+
+Many distributions already provide ways to query for the cluster domain (such as
+kubeadm[^prior-art-kubeadm], k3s[^prior-art-k3s], and OpenShift[^prior-art-openshift]).
+However, these are all inconsistent with each other, requiring applications to
+provide special cases for each.
+
+[^prior-art-kubeadm]: kubeadm (including kind) creates a ConfigMap `kube-system/kubeadm-config` that contains the full kubeadm config, including `networking.dnsDomain`.
+[^prior-art-k3s]: k3s creates a ConfigMap `kube-system/clusterdns` that contains it as the `.data.clusterDomain` field. See <https://github.com/k3s-io/k3s/pull/1785>.
+[^prior-art-openshift]: OpenShift defines a custom DNS CRD that contains it as the `.status.clusterDomain` field. See <https://docs.openshift.com/container-platform/4.17/networking/dns-operator.html#nw-dns-view_dns-operator>.
+
+It can also be retrieved from the kubelet's `/configz` endpoint, however this is
+[considered unstable](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/staging/src/k8s.io/component-base/configz/OWNERS#L3-L5).
+
```

Comment:
IMHO if you depend on that to get the cluster domain field and it changes breaking consumers, I would consider that a regression and we'll fix it.

I don't think that comment means that this endpoint or functionality is going to disappear, it is more about the schema of the config, we have a similar thing with kube-proxy config that is v1alpha1, but at this point and after more than X years it seems to me those are now stable APIs we can not change,.

 cc: @thockin